tools/dotcms-cli/api-data-model/src/main/java/com/dotcms/api/traversal/Filter.java

```diff
@@ -0,0 +1,285 @@
+import com.dotcms.model.asset.AssetView;
+import com.dotcms.model.asset.FolderView;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.PathMatcher;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A filter that determines whether a path should be included or excluded based on the given
+ * includes and excludes patterns. This filter can be applied to folders and assets separately.
+ */
+public class Filter {
+
+    private final Set<PathMatcher> folderIncludes;
+    private final Set<PathMatcher> assetIncludes;
+    private final Set<PathMatcher> folderExcludes;
+    private final Set<PathMatcher> assetExcludes;
+    private final String rootPath;
+
+    /**
+     * Creates a new filter with the specified includes and excludes patterns for folders and
+     * assets, and the root path.
+     *
+     * @param folderIncludes The patterns to include folders.
+     * @param assetIncludes  The patterns to include assets.
+     * @param folderExcludes The patterns to exclude folders.
+     * @param assetExcludes  The patterns to exclude assets.
+     * @param rootPath       The root path to apply the filter. This is used to remove the root path
+     *                       prefix from the folder and asset paths before applying the filter.
+     */
+    private Filter(
+            Set<PathMatcher> folderIncludes,
+            Set<PathMatcher> assetIncludes,
+            Set<PathMatcher> folderExcludes,
+            Set<PathMatcher> assetExcludes,
+            String rootPath) {
+        this.folderIncludes = folderIncludes;
+        this.assetIncludes = assetIncludes;
+        this.folderExcludes = folderExcludes;
+        this.assetExcludes = assetExcludes;
+        this.rootPath = rootPath;
+    }
+
+    /**
+     * Applies the filter to the given folder, filtering its sub-folders and assets based on the
+     * configured includes and excludes patterns.
+     *
+     * @param folder The folder to apply the filter to.
+     * @return The filtered folder view.
+     */
+    public FolderView apply(FolderView folder) {
+
+        List<FolderView> filteredSubFolders = new ArrayList<>();
+        List<AssetView> filteredAssets = new ArrayList<>();
+
+        if (folder.subFolders() != null) {
+            folder.subFolders().iterator().forEachRemaining(subFolder -> {
+
+                subFolder = validateFolder(subFolder);
+                filteredSubFolders.add(subFolder);
+            });
+        }
+
+        if (folder.assets() != null) {
+            folder.assets().versions().iterator().forEachRemaining(assetVersion -> {
+
+                var assetVersionPath = assetVersion.metadata().get(PATH_META_KEY.key()).toString();
+                var assetPath = Paths.get(assetVersionPath).
+                        resolve(assetVersion.name()).toString();
+
+                if (!rootPath.equals("/")) {
```

Comment:
shouldn't be this O.S compatible?

Reply:
Not really, as it is used in the context of URLs and not file paths.